In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
import imblearn

In [2]:
df_assess = pd.read_csv("../MSDS_Capstone/assessments.csv")
df_course = pd.read_csv("../MSDS_Capstone/courses.csv")
df_stud_asses = pd.read_csv("../MSDS_Capstone/studentAssessment.csv")
df_stud_info = pd.read_csv("../MSDS_Capstone/studentInfo.csv")
df_stud_reg = pd.read_csv("../MSDS_Capstone/studentRegistration.csv")
df_stud_vle = pd.read_csv("../MSDS_Capstone/studentVle.csv")
df_vle = pd.read_csv("../MSDS_Capstone/vle.csv")


1. Title of Database: Open University Learning Analytics dataset

2. Sources:

   (a) Original owners of database:
  The Open University
  Walton Hall, Milton Keynes
  MK7 6AA
  United Kingdom
  contact: Zdenek Zdrahal (zdenek.zdrahal@open.ac.uk)

   (b) Donor of database:
  Jakub Kuzilek (jakub.kuzilek@gmail.com)
  Knowledge Media Institute, The Open University
  Walton Hall, Milton Keynes, MK7 6AA, UK
  CIIRC, CTU in Prague
  Zikova 1903/4, Prague, CZ
  Martin Hlosta (martin.hlosta@open.ac.uk)
  Knowledge Media Institute, The Open University
  Walton Hall, Milton Keynes, MK7 6AA, UK
  Zdenek Zdrahal (zdenek.zdrahal@open.ac.uk)
  Knowledge Media Institute, The Open University
  Walton Hall, Milton Keynes, MK7 6AA, UK
  CIIRC, CTU in Prague
  Zikova 1903/4, Prague, CZ

   (c) Date received: December 2015


3. Past Usage:
  - none -

4. Relevant Information Paragraph:

    Open University Learning Analytics Dataset (OULAD) contains data about 
    courses, students and their interactions with Virtual Learning 
    Environment (VLE) for seven selected courses (called modules). Presentations
    of courses start in February and October - they are marked by "B" and "J" 
    respectively. The dataset consists of tables connected using unique 
    identifiers. Dataset is stored in several csv files.  
    
    More information, examples and news can be found at: 
      https://analyse.kmi.open.ac.uk/open_dataset

5. Number of Instances: 
    - students in courses: 32953
    - course-presentations: 22
    - VLE pages: 6364
    - VLE log entries: 10655280
    - registration entries: 32953
    - assessments: 206
    - assessment entries: 173912


6. Number of Attributes: 43


7. Attribute information:
   Name		              Data Type	  Meas.	 Description
	 ----                 ---------   -----  -------------------------------------
   -----------------------------------------------------------------------------
   courses.csv
   -----------------------------------------------------------------------------
   code_module          nominal            code name of the module, which serves 
                                           as the identifier
   code_presentation    nominal            code name of the presentation. It 
                                           consists of the year and  B  for the 
                                           presentation starting in February and 
                                            J  for the presentation starting 
                                           in October
   length               integer     days   length of the module-presentation 
                                           in days
   -----------------------------------------------------------------------------                                      
   assessments.csv
   -----------------------------------------------------------------------------
   code_module          nominal            identification code of the module, 
                                           to which the assessment belongs
   code_presentation    nominal            identification code of the 
                                           presentation, to which the assessment 
                                           belongs
   id_assessment        integer            identification number of the 
                                           assessment
   assessment_type      nominal            type of assessment
   date                 integer     days   information about the final 
                                           submission date of the assessment 
                                           calculated as the number of days 
                                           since the start of the 
                                           module-presentation. The starting 
                                           date of the presentation has 
                                           number 0 (zero)
   weight               integer    %       weight of the assessment in %. 
                                           Typically, Exams are treated separa-
                                           tely and have the weight 100%; 
                                           the sum of all other assessments 
                                           is 100%
   -----------------------------------------------------------------------------                                      
   vle.csv
   -----------------------------------------------------------------------------
   id_site              integer            an identification number 
                                           of the material
   code_module          nominal            an identification code for module
   code_presentation    nominal            the identification code of presenta-
                                           tion
   activity_type        nominal            the role associated with the module 
                                           material
   week_from            integer    week    the week from which the material 
                                           is planned to be used
   week_to              integer    week    week until which the material is 
                                           planned to be used
   -----------------------------------------------------------------------------
   studentInfo.csv
   -----------------------------------------------------------------------------
   code_module          nominal            an identification code for a module 
                                           on which the student is registered
   code_presentation    nominal            the identification code of the 
                                           presentation during which the student 
                                           is registered on the module
   id_student           integer            a unique identification number 
                                           for the student
   gender               nominal            the student s gender
   region               nominal            identifies the geographic region, 
                                           where the student lived while taking 
                                           the module-presentation
   highest_education    nominal            highest student education level 
                                           on entry to the module presentation
   imd_band             nominal            specifies the Index of Multiple 
                                           Depravation band of the place where 
                                           the student lived during 
                                           the module-presentation
   age_band             nominal            band of the student s age
   num_of_prev_attempts integer            the number times the student has 
                                           attempted this module
   studied_credits      integer            the total number of credits for the 
                                           modules the student is currently 
                                           studying
   disability           nominal            indicates whether the student has 
                                           declared a disability
   final_result         nominal            student s final result in the 
                                           module-presentation
   -----------------------------------------------------------------------------                                        
   studentRegistration.csv
   -----------------------------------------------------------------------------
   code_module          nominal            an identification code for a module
   code_presentation    nominal            the identification code 
                                           of the presentation
   id_student           integer            a unique identification number 
                                           for the student
   date_registration    integer    days    the date of student s registration on
                                           the module presentation, this is the 
                                           number of days measured relative to 
                                           the start of the module-presentation 
                                           (e.g. the negative value -30 means 
                                           that the student registered to module
                                           presentation 30 days before 
                                           it started)
   date_unregistration  integer    days    date of student unregistration from 
                                           the module presentation, this is the 
                                           number of days measured relative to 
                                           the start of the module-presentation.
                                           Students, who completed the course 
                                           have this field empty. Students who 
                                           unregistered have Withdrawal as the 
                                           value of the final_result column in 
                                           the studentInfo.csv file
   -----------------------------------------------------------------------------                                       
   studentAssessment.csv
   -----------------------------------------------------------------------------
   id_assessment        integer            the identification number 
                                           of the assessment
   id_student           integer            a unique identification number 
                                           for the student
   date_submitted       integer    days    the date of student submission, 
                                           measured as the number of days since 
                                           the start of the module presentation
   is_banked            integer            a status flag indicating that the 
                                           assessment result has been 
                                           transferred from a previous 
                                           presentation
   score                integer            the student s score in 
                                           this assessment. The range is from
                                           0 to 100. The score lower than 40 is 
                                           interpreted as Fail. The marks are in
                                           the range from 0 to 100
   -----------------------------------------------------------------------------
   studentVle.csv
   -----------------------------------------------------------------------------
   code_module          nominal            an identification code for a module
   code_presentation    nominal            the identification code of the 
                                           module presentation
   id_student           integer            a unique identification number 
                                           for the student
   id_site              integer            an identification number for the 
                                           VLE material
   date                 integer    days    the date of student s interaction 
                                           with the material measured as the 
                                           number of days since the start of 
                                           the module-presentation
   sum_click            integer            the number of times a student 
                                           interacts with the material in 
                                           that day 

8. Missing Attribute Values: Yes


9. Class Distribution:

	Class	      Examples
	----------- ---------
      Distinction 3024
      ----------- ---------
      Fail        7052 
      ----------- ---------
      Pass        12361
      ----------- ---------
      Withdrawn   10156
      ----------- ---------
      Total       32593

In [3]:
print("Assessments \n",df_assess.head())
print("Assessments \n",df_assess.info())
for col in df_assess.columns:
    print(col)
    print(df_assess[col].unique())
    print(df_assess[col].value_counts())

Assessments 
   code_module code_presentation  id_assessment assessment_type date  weight
0         AAA             2013J           1752             TMA   19    10.0
1         AAA             2013J           1753             TMA   54    20.0
2         AAA             2013J           1754             TMA  117    20.0
3         AAA             2013J           1755             TMA  166    20.0
4         AAA             2013J           1756             TMA  215    30.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code_module        206 non-null    object 
 1   code_presentation  206 non-null    object 
 2   id_assessment      206 non-null    int64  
 3   assessment_type    206 non-null    object 
 4   date               206 non-null    object 
 5   weight             206 non-null    float64
dtypes: float64(1), int64(1), object(4)
me

This information is the details of the assessments. 
There is TMA - Tutor Marked assessment, CMA, Computer marked assessment and Exam. 
TMA and CMA are scheduled on the course dates. The exam can be any time after the end.
Exam = 100%, TMA & CMA add up to 100% for the course. 
I cannot see a weighting on their website so will assume 50:50 for the overall grade. Unless it can be ascertained from the other tables. 

In [4]:
print("\n Course \n",df_course.head())
print("\n Course \n",df_course.info())
for col in df_course.columns:
    print(col)
    print(df_course[col].unique())


 Course 
   code_module code_presentation  module_presentation_length
0         AAA             2013J                         268
1         AAA             2014J                         269
2         BBB             2013J                         268
3         BBB             2014J                         262
4         BBB             2013B                         240
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   code_module                 22 non-null     object
 1   code_presentation           22 non-null     object
 2   module_presentation_length  22 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 660.0+ bytes

 Course 
 None
code_module
['AAA' 'BBB' 'CCC' 'DDD' 'EEE' 'FFF' 'GGG']
code_presentation
['2013J' '2014J' '2013B' '2014B']
module_presentation_length
[268 269 262 240 234 241 261]


This is just the course information. Important part here could be the length to understand the frequency of the assessments

In [5]:
print("\n Student Assessment \n",df_stud_asses.head())
print("\n Student Assessment \n",df_stud_asses.info())
for col in df_stud_asses.columns:
    print(col)
    print(df_stud_asses[col].unique())
print(df_stud_asses["is_banked"].value_counts())



 Student Assessment 
    id_assessment  id_student  date_submitted  is_banked score
0           1752       11391              18          0    78
1           1752       28400              22          0    70
2           1752       31604              17          0    72
3           1752       32885              26          0    69
4           1752       38053              19          0    79
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 173912 entries, 0 to 173911
Data columns (total 5 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   id_assessment   173912 non-null  int64 
 1   id_student      173912 non-null  int64 
 2   date_submitted  173912 non-null  int64 
 3   is_banked       173912 non-null  int64 
 4   score           173912 non-null  object
dtypes: int64(4), object(1)
memory usage: 6.6+ MB

 Student Assessment 
 None
id_assessment
[ 1752  1753  1754  1755  1756  1758  1759  1760  1761  1762 14984 14985
 14986 14987 14

Here is the measure we are looking for the grades. <40 is a fail. Up to 70 is a first class honours if doing an honours degree. If not the higher the better. 
Need to combine with the assessment table the type of assessment it is to see if TMA / CMA = Exam or is comparable 

In [6]:
print("\n Student Info \n",df_stud_info.head())
print("\n Student Info \n",df_stud_info.info())
for col in df_stud_info.columns:
    print(col)
    print(df_stud_info[col].unique())
    print(df_stud_info[col].value_counts())


 Student Info 
   code_module code_presentation  id_student gender                region  \
0         AAA             2013J       11391      M   East Anglian Region   
1         AAA             2013J       28400      F              Scotland   
2         AAA             2013J       30268      F  North Western Region   
3         AAA             2013J       31604      F     South East Region   
4         AAA             2013J       32885      F  West Midlands Region   

       highest_education imd_band age_band  num_of_prev_attempts  \
0       HE Qualification  90-100%     55<=                     0   
1       HE Qualification   20-30%    35-55                     0   
2  A Level or Equivalent   30-40%    35-55                     0   
3  A Level or Equivalent   50-60%    35-55                     0   
4     Lower Than A Level   50-60%     0-35                     0   

   studied_credits disability final_result  
0              240          N         Pass  
1               60         

A lot of information here to be used. ID is unique, course is the course so a join field.
Gender likely an indicator field, region could be, highest education prior I am interested in looking at, imd is a score of how deprived an area is. The lower the worse off the area is. age is grouped under 35, 35-55 and greater than 55. most fall in the under 35 category with a small percentage over 55. May be an interesting stat but could do with more granularity. 
previous attempts suggest that most only have 1 attempt with resits declining the more attempts. studied credits is just progress. pass withdrawn or fail for the final course resut will tie to assessments. May pass some but not all. could be an at risk indicator. 

In [7]:
print("\n Student Registration \n",df_stud_reg.head())
print("\n Student Registration \n",df_stud_reg.info())
for col in df_stud_reg.columns:
    print(col)
    print(df_stud_reg[col].unique())
    print(df_stud_reg[col].value_counts())


 Student Registration 
   code_module code_presentation  id_student date_registration  \
0         AAA             2013J       11391              -159   
1         AAA             2013J       28400               -53   
2         AAA             2013J       30268               -92   
3         AAA             2013J       31604               -52   
4         AAA             2013J       32885              -176   

  date_unregistration  
0                   ?  
1                   ?  
2                  12  
3                   ?  
4                   ?  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   code_module          32593 non-null  object
 1   code_presentation    32593 non-null  object
 2   id_student           32593 non-null  int64 
 3   date_registration    32593 non-null  object
 4   date_unregistration  32593 non-nu

when the students registered, unregistered means they dropped the course so not a missing value but could use a 999 or something else. The days when they registered may have an impact. will have to investigate. 

In [8]:

print("\n Student VLE \n",df_stud_vle.head())
print("\n Student VLE \n",df_stud_vle.info())

for col in df_stud_vle.columns:
    print(col)
    print(df_stud_vle[col].unique())


 Student VLE 
   code_module code_presentation  id_student  id_site  date  sum_click
0         AAA             2013J       28400   546652   -10          4
1         AAA             2013J       28400   546652   -10          1
2         AAA             2013J       28400   546652   -10          1
3         AAA             2013J       28400   546614   -10         11
4         AAA             2013J       28400   546714   -10          1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10655280 entries, 0 to 10655279
Data columns (total 6 columns):
 #   Column             Dtype 
---  ------             ----- 
 0   code_module        object
 1   code_presentation  object
 2   id_student         int64 
 3   id_site            int64 
 4   date               int64 
 5   sum_click          int64 
dtypes: int64(4), object(2)
memory usage: 487.8+ MB

 Student VLE 
 None
code_module
['AAA' 'BBB' 'CCC' 'DDD' 'EEE' 'FFF' 'GGG']
code_presentation
['2013J' '2014J' '2013B' '2014B']
id_student
[ 28400  3

This is What is being investigated. When the students accessed different areas and how many clicks. Will need to add the VLE infor and aggregate on a few different metrics to see this. 
Longest course is 269 days. This looks to be the highest number so it was not accessed after the course ended potentially. [it is done so why would it be]. 
does the element mean more clicks?

In [9]:
print("\n VLE \n",df_vle.head())
print("\n VLE \n",df_vle.info())
for col in df_vle.columns:
    print(col)
    print(df_vle[col].unique())


 VLE 
    id_site code_module code_presentation activity_type week_from week_to
0   546943         AAA             2013J      resource         ?       ?
1   546712         AAA             2013J     oucontent         ?       ?
2   546998         AAA             2013J      resource         ?       ?
3   546888         AAA             2013J           url         ?       ?
4   547035         AAA             2013J      resource         ?       ?
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6364 entries, 0 to 6363
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id_site            6364 non-null   int64 
 1   code_module        6364 non-null   object
 2   code_presentation  6364 non-null   object
 3   activity_type      6364 non-null   object
 4   week_from          6364 non-null   object
 5   week_to            6364 non-null   object
dtypes: int64(1), object(5)
memory usage: 298.4+ KB

 VLE 
 None
id_site

indicator to the student VLE letting me know what each element is. This will be added to the data to see what gets the most clicks. 

# To Do
- [ ] combine the data basic VLE
<details>
    Combine all interactions and then determine the pass fail for student df_stud_info is all that is required
</details>
- [ ] combine the data student exams
<details>
    Combine exams to a single point of pass fail to assess if interactions impact this part.
</details>
- [ ] combine the data intermediate VLE
<details>
    Combine interactions basaed on the element of the site accessed and look at correlation to pass / fail in different areas.
</details>



In [10]:
df_vle_info = pd.merge(df_stud_vle,df_vle, on = "id_site",copy = False)

In [11]:
df_vle_info.head()

code_module_x code_presentation_x  id_student  id_site  date  sum_click  \
0           AAA               2013J       28400   546652   -10          4   
1           AAA               2013J       28400   546652   -10          1   
2           AAA               2013J       28400   546652   -10          1   
3           AAA               2013J       28400   546652   -10          8   
4           AAA               2013J       30268   546652   -10          3   

  code_module_y code_presentation_y activity_type week_from week_to  
0           AAA               2013J       forumng         ?       ?  
1           AAA               2013J       forumng         ?       ?  
2           AAA               2013J       forumng         ?       ?  
3           AAA               2013J       forumng         ?       ?  
4           AAA               2013J       forumng         ?       ?

In [12]:

df_vle_info.rename(columns = {"code_module_x":"code_module","code_presentation_x":"code_presentation"},inplace = True)

In [13]:
print(df_vle_info.columns)

Index(['code_module', 'code_presentation', 'id_student', 'id_site', 'date',
       'sum_click', 'code_module_y', 'code_presentation_y', 'activity_type',
       'week_from', 'week_to'],
      dtype='object')


In [14]:
df_vle_info = df_vle_info.drop(["code_module_y", "code_presentation_y"],axis=1)


In [15]:
df_vle_info.head()

code_module code_presentation  id_student  id_site  date  sum_click  \
0         AAA             2013J       28400   546652   -10          4   
1         AAA             2013J       28400   546652   -10          1   
2         AAA             2013J       28400   546652   -10          1   
3         AAA             2013J       28400   546652   -10          8   
4         AAA             2013J       30268   546652   -10          3   

  activity_type week_from week_to  
0       forumng         ?       ?  
1       forumng         ?       ?  
2       forumng         ?       ?  
3       forumng         ?       ?  
4       forumng         ?       ?

In [16]:
df_vle_grouped = df_vle_info.groupby(["id_student","code_module","code_presentation"]).sum(numeric_only = True)
#.agg({'activity_type': 'first', 'sum_click':'sum'})

In [17]:
df_vle_grouped = df_vle_grouped.drop(["id_site","date"],axis=1)

In [18]:
df_vle_grouped.head()

sum_click
id_student code_module code_presentation           
6516       AAA         2014J                   2791
8462       DDD         2013J                    646
                       2014J                     10
11391      AAA         2013J                    934
23629      BBB         2013B                    161

In [19]:
df_stud_grp = pd.merge(df_stud_asses,df_assess, on = "id_assessment")

In [20]:
df_stud_grp.head()

id_assessment  id_student  date_submitted  is_banked score code_module  \
0           1752       11391              18          0    78         AAA   
1           1752       28400              22          0    70         AAA   
2           1752       31604              17          0    72         AAA   
3           1752       32885              26          0    69         AAA   
4           1752       38053              19          0    79         AAA   

  code_presentation assessment_type date  weight  
0             2013J             TMA   19    10.0  
1             2013J             TMA   19    10.0  
2             2013J             TMA   19    10.0  
3             2013J             TMA   19    10.0  
4             2013J             TMA   19    10.0

In [21]:
df_stud_grp_2 = pd.merge(df_stud_grp,df_stud_info, on = "id_student")

In [22]:
df_stud_grp_2.head()

id_assessment  id_student  date_submitted  is_banked score code_module_x  \
0           1752       11391              18          0    78           AAA   
1           1753       11391              53          0    85           AAA   
2           1754       11391             115          0    80           AAA   
3           1755       11391             164          0    85           AAA   
4           1756       11391             212          0    82           AAA   

  code_presentation_x assessment_type date  weight  ... code_presentation_y  \
0               2013J             TMA   19    10.0  ...               2013J   
1               2013J             TMA   54    20.0  ...               2013J   
2               2013J             TMA  117    20.0  ...               2013J   
3               2013J             TMA  166    20.0  ...               2013J   
4               2013J             TMA  215    30.0  ...               2013J   

  gender               region highest_education imd_band age_band  \
0      M  East Anglian Region  HE Qualification  90-100%     55<=   
1      M  East Anglian Region  HE Qualification  90-100%     55<=   
2      M  East Anglian Region  HE Qualification  90-100%     55<=   
3      M  East Anglian Region  HE Qualification  90-100%     55<=   
4      M  East Anglian Region  HE Qualification  90-100%     55<=   

  num_of_prev_attempts  studied_credits  disability final_result  
0                    0              240           N         Pass  
1                    0              240           N         Pass  
2                    0              240           N         Pass  
3                    0              240           N         Pass  
4                    0              240           N         Pass  

[5 rows x 21 columns]

In [23]:
df_stud_grp_2.columns

Index(['id_assessment', 'id_student', 'date_submitted', 'is_banked', 'score',
       'code_module_x', 'code_presentation_x', 'assessment_type', 'date',
       'weight', 'code_module_y', 'code_presentation_y', 'gender', 'region',
       'highest_education', 'imd_band', 'age_band', 'num_of_prev_attempts',
       'studied_credits', 'disability', 'final_result'],
      dtype='object')

In [24]:
df_1st_run = pd.merge(df_vle_grouped,df_stud_info, on = ["id_student","code_module","code_presentation"])

In [25]:
df_1st_run.head(30)

id_student code_module code_presentation  sum_click gender  \
0         6516         AAA             2014J       2791      M   
1         8462         DDD             2013J        646      M   
2         8462         DDD             2014J         10      M   
3        11391         AAA             2013J        934      M   
4        23629         BBB             2013B        161      F   
5        23698         CCC             2014J        910      F   
6        23798         BBB             2013J        590      M   
7        24186         GGG             2014B        184      F   
8        24213         DDD             2014B       1992      F   
9        24391         GGG             2013J        712      M   
10       24734         AAA             2014J        499      F   
11       25107         BBB             2013B       2562      F   
12       25150         EEE             2014B       2103      M   
13       25261         CCC             2014J        997      F   
14       25572         DDD             2014J        113      F   
15       25629         BBB             2014B         16      F   
16       25997         BBB             2014B         13      F   
17       26023         GGG             2013J        786      M   
18       26192         AAA             2014J       2223      F   
19       26211         FFF             2014B      15205      F   
20       26247         FFF             2013J        478      M   
21       26315         BBB             2014J       2451      F   
22       26734         BBB             2014J        210      M   
23       26915         EEE             2013J         61      M   
24       27116         CCC             2014J       1847      M   
25       27189         FFF             2014B       2638      F   
26       27417         DDD             2013J        410      M   
27       27417         DDD             2014J        100      M   
28       27683         FFF             2014B       2129      M   
29       27759         BBB             2013J        366      M   

                  region      highest_education imd_band age_band  \
0               Scotland       HE Qualification   80-90%     55<=   
1          London Region       HE Qualification   30-40%     55<=   
2          London Region       HE Qualification   30-40%     55<=   
3    East Anglian Region       HE Qualification  90-100%     55<=   
4    East Anglian Region     Lower Than A Level   20-30%     0-35   
5    East Anglian Region  A Level or Equivalent   50-60%     0-35   
6                  Wales  A Level or Equivalent   50-60%     0-35   
7       Yorkshire Region     Lower Than A Level    10-20     0-35   
8    East Anglian Region  A Level or Equivalent   40-50%     0-35   
9   East Midlands Region  A Level or Equivalent   80-90%     0-35   
10          South Region     Lower Than A Level   60-70%     0-35   
11   East Anglian Region     Lower Than A Level   20-30%     0-35   
12  North Western Region       HE Qualification  90-100%     0-35   
13              Scotland       HE Qualification   40-50%     0-35   
14         London Region       HE Qualification    10-20     0-35   
15              Scotland     Lower Than A Level   20-30%     0-35   
16         London Region  A Level or Equivalent   20-30%     0-35   
17  North Western Region     Lower Than A Level    10-20    35-55   
18   East Anglian Region  A Level or Equivalent   60-70%    35-55   
19         London Region       HE Qualification   20-30%    35-55   
20     South East Region       HE Qualification   80-90%    35-55   
21   East Anglian Region     Lower Than A Level   60-70%    35-55   
22     South West Region  A Level or Equivalent   20-30%     0-35   
23                 Wales     Lower Than A Level    0-10%    35-55   
24      Yorkshire Region       HE Qualification   70-80%     0-35   
25          South Region     Lower Than A Level    10-20    35-55   
26     South West Region     Lower Than A Level   60-70%     0-35   
27     South West Region 

In [26]:
df_1st_run.shape

(29228, 13)

df_1st_run.to_csv("summed_total.csv",index = False)

In [27]:
df_stud_80868 = df_stud_vle.loc[df_stud_vle['id_student'] == 80868]

df_stud_80868.to_csv("df_stud_80868.csv",index = False)

In [28]:
df_vle_grouped2 = df_vle_info.groupby(["id_student","code_module","code_presentation","activity_type"]).agg({'id_site': 'first', 'sum_click':'sum'})
#

In [29]:
df_vle_grouped2

id_site  sum_click
id_student code_module code_presentation activity_type                    
6516       AAA         2014J             dataplus        877423         21
                                         forumng         877025        451
                                         homepage        877030        497
                                         oucontent       877053       1505
                                         resource        877211         31
...                                                         ...        ...
2698588    BBB         2014J             oucollaborate   913734          7
                                         oucontent       913528        233
                                         resource        913680         49
                                         subpage         913696         76
                                         url             913740          2

[240357 rows x 2 columns]

In [30]:
df_ds2 = pd.merge(df_vle_grouped2,df_stud_info, on = ["id_student","code_module","code_presentation"])

In [31]:
df_ds2

id_student code_module code_presentation  id_site  sum_click gender  \
0             6516         AAA             2014J   877423         21      M   
1             6516         AAA             2014J   877025        451      M   
2             6516         AAA             2014J   877030        497      M   
3             6516         AAA             2014J   877053       1505      M   
4             6516         AAA             2014J   877211         31      M   
...            ...         ...               ...      ...        ...    ...   
240352     2698588         BBB             2014J   913734          7      F   
240353     2698588         BBB             2014J   913528        233      F   
240354     2698588         BBB             2014J   913680         49      F   
240355     2698588         BBB             2014J   913696         76      F   
240356     2698588         BBB             2014J   913740          2      F   

                  region highest_education imd_band age_band  \
0               Scotland  HE Qualification   80-90%     55<=   
1               Scotland  HE Qualification   80-90%     55<=   
2               Scotland  HE Qualification   80-90%     55<=   
3               Scotland  HE Qualification   80-90%     55<=   
4               Scotland  HE Qualification   80-90%     55<=   
...                  ...               ...      ...      ...   
240352  Yorkshire Region  HE Qualification   30-40%     0-35   
240353  Yorkshire Region  HE Qualification   30-40%     0-35   
240354  Yorkshire Region  HE Qualification   30-40%     0-35   
240355  Yorkshire Region  HE Qualification   30-40%     0-35   
240356  Yorkshire Region  HE Qualification   30-40%     0-35   

        num_of_prev_attempts  studied_credits disability final_result  
0                          0               60          N         Pass  
1                          0               60          N         Pass  
2                          0               60          N         Pass  
3                          0               60          N         Pass  
4                          0               60          N         Pass  
...                      ...              ...        ...          ...  
240352                     0              120          N  Distinction  
240353                     0              120          N  Distinction  
240354                     0              120          N  Distinction  
240355                     0              120          N  Distinction  
240356                     0              120          N  Distinction  

[240357 rows x 14 columns]

In [32]:
df_ds2 = pd.merge(df_ds2,df_vle[['id_site','activity_type']],on='id_site', how='left')

In [33]:
df_ds2

id_student code_module code_presentation  id_site  sum_click gender  \
0             6516         AAA             2014J   877423         21      M   
1             6516         AAA             2014J   877025        451      M   
2             6516         AAA             2014J   877030        497      M   
3             6516         AAA             2014J   877053       1505      M   
4             6516         AAA             2014J   877211         31      M   
...            ...         ...               ...      ...        ...    ...   
240352     2698588         BBB             2014J   913734          7      F   
240353     2698588         BBB             2014J   913528        233      F   
240354     2698588         BBB             2014J   913680         49      F   
240355     2698588         BBB             2014J   913696         76      F   
240356     2698588         BBB             2014J   913740          2      F   

                  region highest_education imd_band age_band  \
0               Scotland  HE Qualification   80-90%     55<=   
1               Scotland  HE Qualification   80-90%     55<=   
2               Scotland  HE Qualification   80-90%     55<=   
3               Scotland  HE Qualification   80-90%     55<=   
4               Scotland  HE Qualification   80-90%     55<=   
...                  ...               ...      ...      ...   
240352  Yorkshire Region  HE Qualification   30-40%     0-35   
240353  Yorkshire Region  HE Qualification   30-40%     0-35   
240354  Yorkshire Region  HE Qualification   30-40%     0-35   
240355  Yorkshire Region  HE Qualification   30-40%     0-35   
240356  Yorkshire Region  HE Qualification   30-40%     0-35   

        num_of_prev_attempts  studied_credits disability final_result  \
0                          0               60          N         Pass   
1                          0               60          N         Pass   
2                          0               60          N         Pass   
3                          0               60          N         Pass   
4                          0               60          N         Pass   
...                      ...              ...        ...          ...   
240352                     0              120          N  Distinction   
240353                     0              120          N  Distinction   
240354                     0              120          N  Distinction   
240355                     0              120          N  Distinction   
240356                     0              120          N  Distinction   

        activity_type  
0            dataplus  
1             forumng  
2            homepage  
3           oucontent  
4            resource  
...               ...  
240352  oucollaborate  
240353      oucontent  
240354       resource  
240355        subpage  
240356            url  

[240357 rows x 15 columns]

In [34]:
len(pd.unique(df_ds2['id_student']))

26074

df_ds2.to_csv("DS2_graph.csv",index = False)

In [35]:
df_ds2 = df_ds2.pivot(index=["id_student","code_module","code_presentation"],columns="activity_type",values = "sum_click")

In [36]:
df_ds2.replace(np.nan, 0, inplace=True)

In [37]:
df_ds2

activity_type                             dataplus  dualpane  externalquiz  \
id_student code_module code_presentation                                     
6516       AAA         2014J                  21.0       0.0           0.0   
8462       DDD         2013J                   0.0       0.0          12.0   
                       2014J                   0.0       0.0           0.0   
11391      AAA         2013J                   0.0       0.0           0.0   
23629      BBB         2013B                   0.0       0.0           0.0   
...                                            ...       ...           ...   
2698257    AAA         2013J                   0.0       0.0           0.0   
2698535    CCC         2014B                   0.0       0.0           0.0   
           EEE         2013J                   0.0       2.0           0.0   
2698577    BBB         2014J                   0.0       0.0           0.0   
2698588    BBB         2014J                   0.0       0.0           0.0   

activity_type                             folder  forumng  glossary  homepage  \
id_student code_module code_presentation                                        
6516       AAA         2014J                 0.0    451.0       0.0     497.0   
8462       DDD         2013J                 0.0     36.0       0.0     184.0   
                       2014J                 0.0      2.0       0.0       7.0   
11391      AAA         2013J                 0.0    193.0       0.0     138.0   
23629      BBB         2013B                 0.0     87.0       0.0      36.0   
...                                          ...      ...       ...       ...   
2698257    AAA         2013J                 0.0    201.0       0.0     165.0   
2698535    CCC         2014B                 0.0    309.0       0.0      98.0   
           EEE         2013J                 0.0   1953.0       0.0     391.0   
2698577    BBB         2014J                 0.0    103.0       0.0     204.0   
2698588    BBB         2014J                 0.0     67.0       0.0     171.0   

activity_type                             htmlactivity  oucollaborate  \
id_student code_module code_presentation                                
6516       AAA         2014J                       0.0            0.0   
8462       DDD         2013J                       0.0           12.0   
                       2014J                       0.0            1.0   
11391      AAA         2013J                       0.0            0.0   
23629      BBB         2013B                       0.0            0.0   
...                                                ...            ...   
2698257    AAA         2013J                       0.0            0.0   
2698535    CCC         2014B                       0.0            0.0   
           EEE         2013J                       0.0            0.0   
2698577    BBB         2014J                       0.0            5.0   
2698588    BBB         2014J                       0.0            7.0   

activity_type                             oucontent  ouelluminate  ouwiki  \
id_student code_module code_presentation                                    
6516       AAA         2014J                 1505.0           0.0     0.0   
8462       DDD         2013J                   64.0           0.0    18.0   
                       2014J                    0.0           0.0     0.0   
11391      AAA         2013J                  553.0           0.0     0.0   
23629      BBB         2013B                    0.0           0.0     0.0   
...                                             ...           ...     ...   
2698257    AAA         2013J                  331.0           0.0     0.0   
2698535    CCC         2014B                    6.0           0.0     0.0   
           EEE         2013J                  773.0           0.0   119.0   
2698577    BBB         2014J                  253.0           0.0     0.0   
2698588    BBB         2014J                  233.0      

In [38]:
df_ds2 = pd.merge(df_ds2,df_stud_info, on = ["id_student","code_module","code_presentation"])

In [39]:
df_ds2

id_student code_module code_presentation  dataplus  dualpane  \
0            6516         AAA             2014J      21.0       0.0   
1            8462         DDD             2013J       0.0       0.0   
2            8462         DDD             2014J       0.0       0.0   
3           11391         AAA             2013J       0.0       0.0   
4           23629         BBB             2013B       0.0       0.0   
...           ...         ...               ...       ...       ...   
29223     2698257         AAA             2013J       0.0       0.0   
29224     2698535         CCC             2014B       0.0       0.0   
29225     2698535         EEE             2013J       0.0       2.0   
29226     2698577         BBB             2014J       0.0       0.0   
29227     2698588         BBB             2014J       0.0       0.0   

       externalquiz  folder  forumng  glossary  homepage  ...    url  gender  \
0               0.0     0.0    451.0       0.0     497.0  ...  143.0       M   
1              12.0     0.0     36.0       0.0     184.0  ...   23.0       M   
2               0.0     0.0      2.0       0.0       7.0  ...    0.0       M   
3               0.0     0.0    193.0       0.0     138.0  ...    5.0       M   
4               0.0     0.0     87.0       0.0      36.0  ...    0.0       F   
...             ...     ...      ...       ...       ...  ...    ...     ...   
29223           0.0     0.0    201.0       0.0     165.0  ...    5.0       M   
29224           0.0     0.0    309.0       0.0      98.0  ...    6.0       M   
29225           0.0     0.0   1953.0       0.0     391.0  ...   90.0       M   
29226           0.0     0.0    103.0       0.0     204.0  ...    4.0       F   
29227           0.0     0.0     67.0       0.0     171.0  ...    2.0       F   

                     region   highest_education  imd_band  age_band  \
0                  Scotland    HE Qualification    80-90%      55<=   
1             London Region    HE Qualification    30-40%      55<=   
2             London Region    HE Qualification    30-40%      55<=   
3       East Anglian Region    HE Qualification   90-100%      55<=   
4       East Anglian Region  Lower Than A Level    20-30%      0-35   
...                     ...                 ...       ...       ...   
29223  East Midlands Region  Lower Than A Level    60-70%      0-35   
29224                 Wales  Lower Than A Level    50-60%      0-35   
29225                 Wales  Lower Than A Level    50-60%      0-35   
29226                 Wales  Lower Than A Level    50-60%     35-55   
29227      Yorkshire Region    HE Qualification    30-40%      0-35   

       num_of_prev_attempts  studied_credits  disability  final_result  
0                         0               60           N          Pass  
1                         0               90           N     Withdrawn  
2                         1               60           N     Withdrawn  
3                         0              240           N          Pass  
4                         2               60           N          Fail  
...                     ...              ...         ...           ...  
29223                     0              120           N          Pass  
29224                     0               60           N     Withdrawn  
29225                     0               60           N          Pass  
29226                     0               60           N          Fail  
29227                     0              120           N   Distinction  

[29228 rows x 32 columns]

df_ds2.to_csv("DS2.csv",index = False)

In [40]:
df_stud_asses

id_assessment  id_student  date_submitted  is_banked score
0                1752       11391              18          0    78
1                1752       28400              22          0    70
2                1752       31604              17          0    72
3                1752       32885              26          0    69
4                1752       38053              19          0    79
...               ...         ...             ...        ...   ...
173907          37443      527538             227          0    60
173908          37443      534672             229          0   100
173909          37443      546286             215          0    80
173910          37443      546724             230          0   100
173911          37443      558486             224          0    80

[173912 rows x 5 columns]

In [41]:
df_assess

code_module code_presentation  id_assessment assessment_type date  weight
0           AAA             2013J           1752             TMA   19    10.0
1           AAA             2013J           1753             TMA   54    20.0
2           AAA             2013J           1754             TMA  117    20.0
3           AAA             2013J           1755             TMA  166    20.0
4           AAA             2013J           1756             TMA  215    30.0
..          ...               ...            ...             ...  ...     ...
201         GGG             2014J          37443             CMA  229     0.0
202         GGG             2014J          37435             TMA   61     0.0
203         GGG             2014J          37436             TMA  124     0.0
204         GGG             2014J          37437             TMA  173     0.0
205         GGG             2014J          37444            Exam  229   100.0

[206 rows x 6 columns]

In [42]:
df_assess["weight"].value_counts()

weight
0.0      56
100.0    24
25.0     15
1.0      15
18.0     15
20.0     13
12.5     11
28.0      9
10.0      8
22.0      6
8.0       4
5.0       4
2.0       3
3.0       3
15.0      3
16.0      3
7.0       2
35.0      2
4.0       2
17.5      2
30.0      2
9.0       2
6.0       1
7.5       1
Name: count, dtype: int64

In [43]:
df_assess_copy = df_assess.copy()

In [44]:
df_assess_copy['assess_unique'] = df_assess_copy[['code_module', 'code_presentation', "assessment_type","date"]].agg('-'.join, axis=1)

In [45]:
df_assess_copy['assess_unique'] = df_assess_copy['assess_unique']+"-"+df_assess_copy['weight'].astype(str)

In [46]:
df_assess_copy

code_module code_presentation  id_assessment assessment_type date  weight  \
0           AAA             2013J           1752             TMA   19    10.0   
1           AAA             2013J           1753             TMA   54    20.0   
2           AAA             2013J           1754             TMA  117    20.0   
3           AAA             2013J           1755             TMA  166    20.0   
4           AAA             2013J           1756             TMA  215    30.0   
..          ...               ...            ...             ...  ...     ...   
201         GGG             2014J          37443             CMA  229     0.0   
202         GGG             2014J          37435             TMA   61     0.0   
203         GGG             2014J          37436             TMA  124     0.0   
204         GGG             2014J          37437             TMA  173     0.0   
205         GGG             2014J          37444            Exam  229   100.0   

                assess_unique  
0       AAA-2013J-TMA-19-10.0  
1       AAA-2013J-TMA-54-20.0  
2      AAA-2013J-TMA-117-20.0  
3      AAA-2013J-TMA-166-20.0  
4      AAA-2013J-TMA-215-30.0  
..                        ...  
201     GGG-2014J-CMA-229-0.0  
202      GGG-2014J-TMA-61-0.0  
203     GGG-2014J-TMA-124-0.0  
204     GGG-2014J-TMA-173-0.0  
205  GGG-2014J-Exam-229-100.0  

[206 rows x 7 columns]

In [47]:
df_stud_assess_2 =  pd.merge(df_stud_asses,df_assess_copy,on = "id_assessment")

In [48]:
df_stud_assess_2

id_assessment  id_student  date_submitted  is_banked score  \
0                1752       11391              18          0    78   
1                1752       28400              22          0    70   
2                1752       31604              17          0    72   
3                1752       32885              26          0    69   
4                1752       38053              19          0    79   
...               ...         ...             ...        ...   ...   
173907          37443      527538             227          0    60   
173908          37443      534672             229          0   100   
173909          37443      546286             215          0    80   
173910          37443      546724             230          0   100   
173911          37443      558486             224          0    80   

       code_module code_presentation assessment_type date  weight  \
0              AAA             2013J             TMA   19    10.0   
1              AAA             2013J             TMA   19    10.0   
2              AAA             2013J             TMA   19    10.0   
3              AAA             2013J             TMA   19    10.0   
4              AAA             2013J             TMA   19    10.0   
...            ...               ...             ...  ...     ...   
173907         GGG             2014J             CMA  229     0.0   
173908         GGG             2014J             CMA  229     0.0   
173909         GGG             2014J             CMA  229     0.0   
173910         GGG             2014J             CMA  229     0.0   
173911         GGG             2014J             CMA  229     0.0   

                assess_unique  
0       AAA-2013J-TMA-19-10.0  
1       AAA-2013J-TMA-19-10.0  
2       AAA-2013J-TMA-19-10.0  
3       AAA-2013J-TMA-19-10.0  
4       AAA-2013J-TMA-19-10.0  
...                       ...  
173907  GGG-2014J-CMA-229-0.0  
173908  GGG-2014J-CMA-229-0.0  
173909  GGG-2014J-CMA-229-0.0  
173910  GGG-2014J-CMA-229-0.0  
173911  GGG-2014J-CMA-229-0.0  

[173912 rows x 11 columns]

In [49]:
df_stud_assess_p = df_stud_assess_2.pivot(index=["id_student","code_module","code_presentation"],columns="id_assessment",values = "score")

In [50]:
df_stud_assess_p

id_assessment                            1752  1753  1754  1755  1756  1758   \
id_student code_module code_presentation                                       
6516       AAA         2014J               NaN   NaN   NaN   NaN   NaN    60   
8462       DDD         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
                       2014J               NaN   NaN   NaN   NaN   NaN   NaN   
11391      AAA         2013J                78    85    80    85    82   NaN   
23629      BBB         2013B               NaN   NaN   NaN   NaN   NaN   NaN   
...                                        ...   ...   ...   ...   ...   ...   
2698257    AAA         2013J                58    64    76    67    74   NaN   
2698535    CCC         2014B               NaN   NaN   NaN   NaN   NaN   NaN   
           EEE         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
2698577    BBB         2014J               NaN   NaN   NaN   NaN   NaN   NaN   
2698588    BBB         2014J               NaN   NaN   NaN   NaN   NaN   NaN   

id_assessment                            1759  1760  1761  1762   ... 37433  \
id_student code_module code_presentation                          ...         
6516       AAA         2014J                48    63    61    77  ...   NaN   
8462       DDD         2013J               NaN   NaN   NaN   NaN  ...   NaN   
                       2014J               NaN   NaN   NaN   NaN  ...   NaN   
11391      AAA         2013J               NaN   NaN   NaN   NaN  ...   NaN   
23629      BBB         2013B               NaN   NaN   NaN   NaN  ...   NaN   
...                                        ...   ...   ...   ...  ...   ...   
2698257    AAA         2013J               NaN   NaN   NaN   NaN  ...   NaN   
2698535    CCC         2014B               NaN   NaN   NaN   NaN  ...   NaN   
           EEE         2013J               NaN   NaN   NaN   NaN  ...   NaN   
2698577    BBB         2014J               NaN   NaN   NaN   NaN  ...   NaN   
2698588    BBB         2014J               NaN   NaN   NaN   NaN  ...   NaN   

id_assessment                            37435 37436 37437 37438 37439 37440  \
id_student code_module code_presentation                                       
6516       AAA         2014J               NaN   NaN   NaN   NaN   NaN   NaN   
8462       DDD         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
                       2014J               NaN   NaN   NaN   NaN   NaN   NaN   
11391      AAA         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
23629      BBB         2013B               NaN   NaN   NaN   NaN   NaN   NaN   
...                                        ...   ...   ...   ...   ...   ...   
2698257    AAA         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
2698535    CCC         2014B               NaN   NaN   NaN   NaN   NaN   NaN   
           EEE         2013J               NaN   NaN   NaN   NaN   NaN   NaN   
2698577    BBB         2014J               NaN   NaN   NaN   NaN   NaN   NaN   
2698588    BBB         2014J               NaN   NaN   NaN   NaN   NaN   NaN   

id_assessment                            37441 37442 37443  
id_student code_module code_presentation                    
6516       AAA         2014J               NaN   NaN   NaN  
8462       DDD         2013J               NaN   NaN   NaN  
                       2014J               NaN   NaN   NaN  
11391      AAA         2013J               NaN   NaN   NaN  
23629      BBB         2013B               NaN   NaN   NaN  
...                                        ...   ...   ...  
2698257    AAA         2013J               NaN   NaN   NaN  
2698535    CCC         2014B               NaN   NaN   NaN  
           EEE         2013J               NaN   NaN   NaN  
2698577    BBB         2014J               NaN   NaN   NaN  
2698588    BBB         2014J               NaN   NaN   NaN  

[25843 rows x 188 columns]

In [51]:
df_stud_assess_2["code_presentation"].value_counts()

code_presentation
2014J    52174
2013J    49018
2014B    38896
2013B    33824
Name: count, dtype: int64

In [52]:
df_stud_assess_p.fillna(0)

id_assessment                            1752  1753  1754  1755  1756  1758   \
id_student code_module code_presentation                                       
6516       AAA         2014J                 0     0     0     0     0    60   
8462       DDD         2013J                 0     0     0     0     0     0   
                       2014J                 0     0     0     0     0     0   
11391      AAA         2013J                78    85    80    85    82     0   
23629      BBB         2013B                 0     0     0     0     0     0   
...                                        ...   ...   ...   ...   ...   ...   
2698257    AAA         2013J                58    64    76    67    74     0   
2698535    CCC         2014B                 0     0     0     0     0     0   
           EEE         2013J                 0     0     0     0     0     0   
2698577    BBB         2014J                 0     0     0     0     0     0   
2698588    BBB         2014J                 0     0     0     0     0     0   

id_assessment                            1759  1760  1761  1762   ... 37433  \
id_student code_module code_presentation                          ...         
6516       AAA         2014J                48    63    61    77  ...     0   
8462       DDD         2013J                 0     0     0     0  ...     0   
                       2014J                 0     0     0     0  ...     0   
11391      AAA         2013J                 0     0     0     0  ...     0   
23629      BBB         2013B                 0     0     0     0  ...     0   
...                                        ...   ...   ...   ...  ...   ...   
2698257    AAA         2013J                 0     0     0     0  ...     0   
2698535    CCC         2014B                 0     0     0     0  ...     0   
           EEE         2013J                 0     0     0     0  ...     0   
2698577    BBB         2014J                 0     0     0     0  ...     0   
2698588    BBB         2014J                 0     0     0     0  ...     0   

id_assessment                            37435 37436 37437 37438 37439 37440  \
id_student code_module code_presentation                                       
6516       AAA         2014J                 0     0     0     0     0     0   
8462       DDD         2013J                 0     0     0     0     0     0   
                       2014J                 0     0     0     0     0     0   
11391      AAA         2013J                 0     0     0     0     0     0   
23629      BBB         2013B                 0     0     0     0     0     0   
...                                        ...   ...   ...   ...   ...   ...   
2698257    AAA         2013J                 0     0     0     0     0     0   
2698535    CCC         2014B                 0     0     0     0     0     0   
           EEE         2013J                 0     0     0     0     0     0   
2698577    BBB         2014J                 0     0     0     0     0     0   
2698588    BBB         2014J                 0     0     0     0     0     0   

id_assessment                            37441 37442 37443  
id_student code_module code_presentation                    
6516       AAA         2014J                 0     0     0  
8462       DDD         2013J                 0     0     0  
                       2014J                 0     0     0  
11391      AAA         2013J                 0     0     0  
23629      BBB         2013B                 0     0     0  
...                                        ...   ...   ...  
2698257    AAA         2013J                 0     0     0  
2698535    CCC         2014B                 0     0     0  
           EEE         2013J                 0     0     0  
2698577    BBB         2014J                 0     0     0  
2698588    BBB         2014J                 0     0     0  

[25843 rows x 188 columns]

df_stud_assess_2.to_csv("df_stud_assess_2.csv")